aten/src/ATen/native/FractionalMaxPool2d.cpp

```diff
@@ -0,0 +1,363 @@
+            AT_ASSERT(w2 >= 0 && w2 < inputW);
+
+            int planeIndex = h2 * inputW + w2;
+            scalar_t val = inputForPlane[planeIndex];
+            if (val > maxVal) {
+              maxVal = val;
+              maxIndex = planeIndex;
+            }
+          }
+        }
+
+        AT_ASSERT(maxVal != -std::numeric_limits<scalar_t>::infinity());
+        AT_ASSERT(maxIndex != -1);
+
+        outputForPlane[h * outputW + w] = maxVal;
+        /* +1 to lua index */
+        indicesForPlane[h * outputW + w] = maxIndex;
+      }
+    }
+
+  }
+}
+
+void fractional_max_pool2d_out_cpu_template(
+  at::Tensor const& input_,
+  at::Tensor& output,
+  IntList output_size,
+  IntList pool_size,
+  at::Tensor& indices,
+  at::Tensor const& randomSamples) {
+
+  int numBatch = 1;
+  int planeDim = 0;
+  int heightDim = 1;
+  int widthDim = 2;
+  int outputW = output_size[0];
+  int outputH = output_size[1];
+  int poolSizeW = pool_size[0];
+  int poolSizeH = pool_size[1];
+
+  /* get contiguous input */
+  auto input = input_.contiguous();
+
+  int ndims = input.ndimension();
+  AT_CHECK((ndims == 3 || ndims == 4),
+		"non-empty 3D or 4D (batch mode) tensor expected for input, but got: ",
+    ndims);
+
+  if (ndims == 4) {
+    numBatch = input.size(0);
+    planeDim++;
+    heightDim++;
+    widthDim++;
+  }
+
+  /* sizes */
+  int numPlanes = input.size(planeDim);
+  int inputH = input.size(heightDim);
+  int inputW = input.size(widthDim);
+
+  AT_CHECK(outputH + poolSizeH - 1 <= inputH,
+    "fractional_max_pool2d(): pool height ", poolSizeH,
+    " too large relative to input height ", inputH);
+  AT_CHECK(outputW + poolSizeW - 1 <= inputW,
+    "fractional_max_pool2d(): pool width ", poolSizeW,
+    " too large relative to input width ", inputW);
+
+  if (ndims == 3) {
+    /* resize output */
+    output.resize_({numPlanes, outputH, outputW});
+    /* indices will contain the locations for each output point */
+    indices.resize_({numPlanes, outputH, outputW});
+
+    AT_DISPATCH_FLOATING_TYPES(input.type(), "fractional_max_pool2d", [&] {
+      auto input_data = input.data<scalar_t>();
+      auto output_data = output.data<scalar_t>();
+      auto indices_data = indices.data<int64_t>();
+      auto randomSamples_data = randomSamples.data<scalar_t>();
+      fractional_max_pool2d_out_frame<scalar_t>(
+        input_data,
+        output_data,
+        indices_data,
+        randomSamples_data,
+        numPlanes, inputW, inputH, outputW, outputH, poolSizeW, poolSizeH);
+      });
+  } else {
+    output.resize_({numBatch, numPlanes, outputH, outputW});
+    /* indices will contain the locations for each output point */
+    indices.resize_({numBatch, numPlanes, outputH, outputW});
+
+    int batch;
+#pragma omp parallel for private(batch)
+    for (batch = 0; batch < numBatch; ++batch) {
+      AT_DISPATCH_FLOATING_TYPES(input.type(), "fractional_max_pool2d", [&] {
+        auto input_data = input.data<scalar_t>();
```

Comment:
nit: Better to do the pointer retrieval outside of the loop; this pointer retrieval is checked and there's no need to do it inside the batch loop. (This probably doesn't matter in practice; it's just the principle of the matter.)

Reply:
Notes on the matter:

```
Many functions in THNN are parallelized in the following way:

    int64_t batch;
#pragma omp parallel for private(batch)
    for (batch = 0; batch < numBatch; ++batch) {
      THNN_(SpatialFractionalMaxPooling_updateOutput_frame)(
        input->data<scalar_t>() + batch * numPlanes * inputH * inputW,
        output->data<scalar_t>() + batch * numPlanes * outputH * outputW,
        THIndexTensor_(data)(indices) + batch * numPlanes * outputH * outputW,
        randomSamples->data<scalar_t>() + batch * numPlanes * 2,
        numPlanes, inputW, inputH, outputW, outputH, poolSizeW, poolSizeH);
    }
  }

This snippet of code is particularly delicate to port, for a few reasons:

1. You need to do a dispatch before you can retrieve the appropriately typed scalar_t pointers, implying use of the AT_DISPATCH_ macro. However...
2. #pragma omp declarations don't work inside lambdas...
3. You do NOT want to do a dynamic dispatch every time around the batch loop; you should do it once to get into the tight loop.

The recommendation is to make a new helper function to contain the OpenMP pragma, and call it from inside the AT_DISPATCH_ macro. Do the extraction of the pointers once, outside of the loop (since they're same every time around.) An example of this being done is in: https://github.com/pytorch/pytorch/pull/15538/files#diff-3e39c997ceefe5db9d4b51504dc8c91cR92 (h/t @Lin Huang)
```

Reply:
Thanks for writing this up!